packages/circuit-ui/components/DateInput/components/DateSegment.tsx

```diff
@@ -0,0 +1,197 @@
+ * limitations under the License.
+ */
+
+'use client';
+
+import {
+  useLayoutEffect,
+  useRef,
+  useState,
+  type ChangeEvent,
+  type InputHTMLAttributes,
+  type KeyboardEvent,
+} from 'react';
+
+import {
+  isArrowLeft,
+  isArrowRight,
+  isBackspace,
+  isDelete,
+} from '../../../util/key-codes.js';
+import { isNumber } from '../../../util/type-check.js';
+import { shiftInRange } from '../../../util/helpers.js';
+import type { DateValue } from '../hooks/usePlainDateState.js';
+import type { SegmentFocus } from '../hooks/useSegmentFocus.js';
+
+import classes from './DateSegment.module.css';
+
+export interface DateSegmentProps
+  extends Omit<
+    InputHTMLAttributes<HTMLInputElement>,
+    'placeholder' | 'value' | 'defaultValue' | 'min' | 'max' | 'onChange'
+  > {
+  placeholder: string;
+  value: DateValue;
+  defaultValue: number;
+  min: number;
+  max: number;
+  step: number;
+  onChange: (value: DateValue) => void;
+  invalid?: boolean;
+  hasWarning?: boolean;
+  showValid?: boolean;
+  readOnly?: boolean;
+  focus: SegmentFocus;
+}
+
+export function DateSegment({
+  onChange,
+  invalid,
+  focus,
+  defaultValue,
+  min,
+  max,
+  step,
+  ...props
+}: DateSegmentProps) {
+  const sizeRef = useRef<HTMLSpanElement>(null);
+  const [width, setWidth] = useState('4ch');
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: The width needs to be recalculated when the value changes
+  useLayoutEffect(() => {
+    if (sizeRef.current) {
+      const cursorWidth = 1;
+      const { offsetWidth } = sizeRef.current;
+      setWidth(`${cursorWidth + offsetWidth}px`);
+    }
+  }, [props.value]);
+
+  const onKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
+    const input = event.currentTarget;
+    const { selectionStart, selectionEnd } = input;
+
+    // Move between segments using arrow keys, but don't interfere with text cursor movement
+    if (selectionStart === selectionEnd) {
+      // Move to the previous segment when the cursor is at the start of the input
+      if (isArrowLeft(event) && (input.readOnly || selectionStart === 0)) {
+        event.preventDefault();
+        focus.previous();
+        return;
+      }
+
+      // Move to the next segment when the cursor is at the end of the input
+      if (
+        isArrowRight(event) &&
+        (input.readOnly || selectionEnd === input.value.length)
+      ) {
+        event.preventDefault();
+        focus.next();
+        return;
+      }
+    }
+
+    // Focus the following segment after clearing the current one
+    if (!input.value) {
+      if (isBackspace(event)) {
+        event.preventDefault();
+        focus.previous();
+        return;
+      }
+
+      if (isDelete(event)) {
+        event.preventDefault();
+        focus.next();
+        return;
+      }
+    }
+
+    // Don't allow editing the value when the input is disabled or read-only
+    if (input.disabled || input.readOnly) {
```

Comment:
why do we need to be explicit about this if the input is disabled ?

Reply:
For some reason, the `onKeyDown` event handler is still called when the `input` disabled. Surprised me as well 😳